plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/router/MikamiTabuchiRouter.java

```diff
@@ -0,0 +1,673 @@
+/*
+ * DBeaver - Universal Database Manager
+ * Copyright (C) 2010-2022 DBeaver Corp and others
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jkiss.dbeaver.erd.ui.router;
+
+
+import org.eclipse.draw2dl.Connection;
+import org.eclipse.draw2dl.IFigure;
+import org.eclipse.draw2dl.geometry.Point;
+import org.eclipse.draw2dl.geometry.PointList;
+import org.eclipse.draw2dl.geometry.PrecisionPoint;
+import org.eclipse.draw2dl.geometry.Rectangle;
+import org.jkiss.code.NotNull;
+import org.jkiss.code.Nullable;
+import org.jkiss.dbeaver.Log;
+import org.jkiss.utils.Pair;
+
+import java.util.*;
+
+/**
+ * Mikami-Tabuchi’s Algorithm
+ * 1. Expand horizontal and vertical line from source to target
+ * 2. In every iteration, expand from the last expanded line by STEP_SIZE
+ * 3. Continue until a line from source intersects another line from target
+ * 4. Backtrace from interception
+ */
+public class MikamiTabuchiRouter {
+
+    private static final Log log = Log.getLog(MikamiTabuchiRouter.class);
+
+    private static final int SOURCE_VERTICAL_LINES = 0;
+    private static final int SOURCE_HORIZONTAL_LINES = 1;
+    private static final int TARGET_VERTICAL_LINES = 2;
+    private static final int TARGET_HORIZONTAL_LINES = 3;
+
+    private int spacing = 15;
+    private final List<Rectangle> obstacles = new ArrayList<>();
+
+    private PrecisionPoint start;
+    private PrecisionPoint finish;
+
+    private final List<OrthogonalPath> workingPaths = new ArrayList<>();
+    private final List<OrthogonalPath> userPaths = new ArrayList<>();
+    private final Map<OrthogonalPath, List<OrthogonalPath>> pathsToChildPaths = new HashMap<>();
+
+    private Set<TrialLine> trialLinesWithPenalty;
+    private Map<Boolean, List<Pair<Point, Point>>> resultMap;
+
+
+    private final Map<Integer, Map<Integer, List<TrialLine>>> linesMap = new LinkedHashMap<>();
```

Comment:
Map can be a bottleneck in a hot code. Consider making it a simple array of four elements. You'll use them all anyway.

```suggestion
    private final Map<Integer, List<TrialLine>[]> linesMap = new LinkedHashMap<>();
```

Also, key key of the `linesMap` is tied to the current iteration. Since it's linear, perhaps the map can once again be replaced with an array/list? Since you will exaust all iterations, you can preallocate it.

```suggestion
    private final TrialLine[][] linesMap = new TrialLine[4][MAX_ITERATIONS_COUNT];
```